File: aurora_monitor/alerts/alert_manager.py
"""
Sistema de Alertas e Notificações.

Gerencia envio de alertas para múltiplos canais:
- Email
- Slack
- Webhooks customizados
- Log
"""

import json
import threading
import time
from collections import deque
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum
from typing import Callable, Dict, List, Optional, Any
import urllib.request
import urllib.error

from aurora_monitor.core.config import AlertConfig


class AlertLevel(Enum):
    """Níveis de alerta."""
    DEBUG = "debug"
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"

    def __lt__(self, other):
        order = [AlertLevel.DEBUG, AlertLevel.INFO, AlertLevel.WARNING,
                 AlertLevel.ERROR, AlertLevel.CRITICAL]
        return order.index(self) < order.index(other)

    def __le__(self, other):
        return self == other or self < other


@dataclass
class Alert:
    """Representa um alerta."""
    level: AlertLevel
    title: str
    message: str
    source: str
    timestamp: datetime = field(default_factory=datetime.now)
    alert_id: str = ""
    tags: Dict[str, str] = field(default_factory=dict)
    metadata: Dict[str, Any] = field(default_factory=dict)
    sent_to: List[str] = field(default_factory=list)
    acknowledged: bool = False
    acknowledged_at: Optional[datetime] = None
    acknowledged_by: Optional[str] = None

    def __post_init__(self):
        if not self.alert_id:
            import hashlib
            content = f"{self.level.value}:{self.title}:{self.source}:{self.timestamp.isoformat()}"
            self.alert_id = hashlib.md5(content.encode()).hexdigest()[:12]

    def to_dict(self) -> Dict[str, Any]:
        """Converte para dicionário."""
        return {
            "alert_id": self.alert_id,
            "level": self.level.value,
            "title": self.title,
            "message": self.message,
            "source": self.source,
            "timestamp": self.timestamp.isoformat(),
            "tags": self.tags,
            "metadata": self.metadata,
            "sent_to": self.sent_to,
            "acknowledged": self.acknowledged,
        }

    def __str__(self) -> str:
        return f"[{self.level.value.upper()}] {self.title}: {self.message}"


@dataclass
class AlertAggregate:
    """Agregação de alertas similares."""
    key: str
    first_occurrence: datetime
    last_occurrence: datetime
    count: int
    alerts: List[Alert] = field(default_factory=list)


class AlertManager:
    """
    Gerenciador de alertas.

    Responsável por:
    - Receber alertas
    - Agregar alertas similares
    - Aplicar cooldown para evitar spam
    - Enviar para canais configurados
    - Manter histórico

    Exemplo de uso:
        manager = AlertManager(config)

        manager.send(
            level=AlertLevel.WARNING,
            title="Alto uso de memória",
            message="Memória em 85%",
            source="aurora.metrics"
        )

        # Com callback
        manager.on_alert(lambda alert: print(alert))
    """

    def __init__(self, config: Optional[AlertConfig] = None):
        """
        Inicializa o gerenciador de alertas.

        Args:
            config: Configuração de alertas.
        """
        self.config = config or AlertConfig()

        self._lock = threading.Lock()
        self._history: deque = deque(maxlen=10000)
        self._last_sent: Dict[str, datetime] = {}
        self._aggregates: Dict[str, AlertAggregate] = {}

        # Callbacks
        self._on_alert_callbacks: List[Callable[[Alert], None]] = []

        # Fila de envio assíncrono
        self._send_queue: deque = deque()
        self._sender_thread: Optional[threading.Thread] = None
        self._sender_running = False

        # Inicia sender se algum canal externo está habilitado
        if any([self.config.email_enabled, self.config.slack_enabled, self.config.webhook_enabled]):
            self._start_sender()

    def _start_sender(self) -> None:
        """Inicia thread de envio assíncrono."""
        self._sender_running = True
        self._sender_thread = threading.Thread(
            target=self._sender_loop,
            name="aurora-alert-sender",
            daemon=True
        )
        self._sender_thread.start()

    def _sender_loop(self) -> None:
        """Loop de envio de alertas."""
        while self._sender_running:
            try:
                if self._send_queue:
                    alert = self._send_queue.popleft()
                    self._send_to_channels(alert)
                else:
                    time.sleep(0.1)
            except Exception:
                time.sleep(1)

    def send(
        self,
        level: AlertLevel,
        title: str,
        message: str,
        source: str,
        tags: Optional[Dict[str, str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Optional[Alert]:
        """
        Envia um alerta.

        Args:
            level: Nível do alerta.
            title: Título do alerta.
            message: Mensagem detalhada.
            source: Origem do alerta.
            tags: Tags para categorização.
            metadata: Metadados adicionais.

        Returns:
            Alert criado ou None se suprimido por cooldown.
        """
        if not self.config.enabled:
            return None

        alert = Alert(
            level=level,
            title=title,
            message=message,
            source=source,
            tags=tags or {},
            metadata=metadata or {},
        )

        # Verifica cooldown
        key = self._get_alert_key(alert)
        if not self._should_send(key):
            self._aggregate(key, alert)
            return None

        with self._lock:
            self._history.append(alert)
            self._last_sent[key] = datetime.now()

        # Notifica callbacks
        self._notify_callbacks(alert)

        # Adiciona à fila de envio
        if any([self.config.email_enabled, self.config.slack_enabled, self.config.webhook_enabled]):
            self._send_queue.append(alert)

        return alert

    def _get_alert_key(self, alert: Alert) -> str:
        """Gera chave única para agregação."""
        return f"{alert.level.value}:{alert.source}:{alert.title}"

    def _should_send(self, key: str) -> bool:
        """Verifica se deve enviar (cooldown)."""
        with self._lock:
            last = self._last_sent.get(key)
            if last is None:
                return True
            elapsed = (datetime.now() - last).total_seconds()
            return elapsed >= self.config.alert_cooldown

    def _aggregate(self, key: str, alert: Alert) -> None:
        """Agrega alerta similar."""
        if not self.config.aggregate_alerts:
            return

        with self._lock:
            if key not in self._aggregates:
                self._aggregates[key] = AlertAggregate(
                    key=key,
                    first_occurrence=alert.timestamp,
                    last_occurrence=alert.timestamp,
                    count=1,
                    alerts=[alert],
                )
            else:
                agg = self._aggregates[key]
                agg.last_occurrence = alert.timestamp
                agg.count += 1
                if len(agg.alerts) < 10:
                    agg.alerts.append(alert)

    def _send_to_channels(self, alert: Alert) -> None:
        """Envia alerta para canais configurados."""
        if self.config.slack_enabled and self.config.slack_webhook_url:
            self._send_to_slack(alert)

        if self.config.webhook_enabled and self.config.custom_webhook_url:
            self._send_to_webhook(alert)

        if self.config.email_enabled and self.config.email_recipients:
            self._send_email(alert)

    def _send_to_slack(self, alert: Alert) -> bool:
        """Envia alerta para Slack."""
        try:
            color_map = {
                AlertLevel.DEBUG: "#808080",
                AlertLevel.INFO: "#36a64f",
                AlertLevel.WARNING: "#ffcc00",
                AlertLevel.ERROR: "#ff6600",
                AlertLevel.CRITICAL: "#ff0000",
            }

            payload = {
                "attachments": [{
                    "color": color_map.get(alert.level, "#808080"),
                    "title": f"[{alert.level.value.upper()}] {alert.title}",
                    "text": alert.message,
                    "fields": [
                        {"title": "Source", "value": alert.source, "short": True},
                        {"title": "Time", "value": alert.timestamp.isoformat(), "short": True},
                    ],
                    "footer": "Aurora Monitor",
                }]
            }

            data = json.dumps(payload).encode('utf-8')
            req = urllib.request.Request(
                self.config.slack_webhook_url,
                data=data,
                headers={'Content-Type': 'application/json'}
            )
            urllib.request.urlopen(req, timeout=10)
            alert.sent_to.append("slack")
            return True

        except Exception:
            return False

    def _send_to_webhook(self, alert: Alert) -> bool:
        """Envia alerta para webhook customizado."""
        try:
            payload = alert.to_dict()
            data = json.dumps(payload).encode('utf-8')
            req = urllib.request.Request(
                self.config.custom_webhook_url,
                data=data,
                headers={'Content-Type': 'application/json'}
            )
            urllib.request.urlopen(req, timeout=10)
            alert.sent_to.append("webhook")
            return True

        except Exception:
            return False

    def _send_email(self, alert: Alert) -> bool:
        """Envia alerta por email."""
        # Implementação básica usando smtplib
        try:
            import smtplib
            from email.mime.text import MIMEText

            msg = MIMEText(f"{alert.message}\n\nSource: {alert.source}\nTime: {alert.timestamp}")
            msg['Subject'] = f"[Aurora] [{alert.level.value.upper()}] {alert.title}"
            msg['From'] = "aurora@localhost"
            msg['To'] = ", ".join(self.config.email_recipients)

            # Requer configuração de SMTP
            # with smtplib.SMTP('localhost') as server:
            #     server.send_message(msg)

            alert.sent_to.append("email")
            return True

        except Exception:
            return False

    def _notify_callbacks(self, alert: Alert) -> None:
        """Notifica callbacks registrados."""
        for callback in self._on_alert_callbacks:
            try:
                callback(alert)
            except Exception:
                pass

    def on_alert(self, callback: Callable[[Alert], None]) -> None:
        """
        Registra callback para alertas.

        Args:
            callback: Função chamada quando alerta é enviado.
        """
        self._on_alert_callbacks.append(callback)

    def acknowledge(self, alert_id: str, by: str = "system") -> bool:
        """
        Reconhece um alerta.

        Args:
            alert_id: ID do alerta.
            by: Quem reconheceu.

        Returns:
            True se alerta foi encontrado e reconhecido.
        """
        with self._lock:
            for alert in self._history:
                if alert.alert_id == alert_id:
                    alert.acknowledged = True
                    alert.acknowledged_at = datetime.now()
                    alert.acknowledged_by = by
                    return True
        return False

    def get_history(
        self,
        limit: int = 100,
        level: Optional[AlertLevel] = None,
        source: Optional[str] = None,
        since: Optional[datetime] = None,
    ) -> List[Alert]:
        """
        Retorna histórico de alertas.

        Args:
            limit: Número máximo de alertas.
            level: Filtrar por nível mínimo.
            source: Filtrar por fonte.
            since: Filtrar por data.

        Returns:
            Lista de alertas.
        """
        with self._lock:
            alerts = list(self._history)

        # Aplica filtros
        if level:
            alerts = [a for a in alerts if a.level >= level]
        if source:
            alerts = [a for a in alerts if source in a.source]
        if since:
            alerts = [a for a in alerts if a.timestamp >= since]

        return alerts[-limit:]

    def get_aggregates(self) -> List[AlertAggregate]:
        """Retorna agregações atuais."""
        with self._lock:
            return list(self._aggregates.values())

    def get_unacknowledged(self, level: Optional[AlertLevel] = None) -> List[Alert]:
        """
        Retorna alertas não reconhecidos.

        Args:
            level: Filtrar por nível mínimo.

        Returns:
            Lista de alertas não reconhecidos.
        """
        with self._lock:
            alerts = [a for a in self._history if not a.acknowledged]

        if level:
            alerts = [a for a in alerts if a.level >= level]

        return alerts

    def get_stats(self) -> Dict[str, Any]:
        """Retorna estatísticas de alertas."""
        with self._lock:
            total = len(self._history)
            by_level = {}
            by_source = {}

            for alert in self._history:
                level = alert.level.value
                by_level[level] = by_level.get(level, 0) + 1

                source = alert.source.split('.')[0]
                by_source[source] = by_source.get(source, 0) + 1

            unack = sum(1 for a in self._history if not a.acknowledged)

            # Últimas 24h
            cutoff = datetime.now() - timedelta(hours=24)
            last_24h = sum(1 for a in self._history if a.timestamp >= cutoff)

        return {
            "enabled": self.config.enabled,
            "total_alerts": total,
            "unacknowledged": unack,
            "last_24_hours": last_24h,
            "by_level": by_level,
            "by_source": by_source,
            "aggregates": len(self._aggregates),
            "queue_size": len(self._send_queue),
            "channels": {
                "email": self.config.email_enabled,
                "slack": self.config.slack_enabled,
                "webhook": self.config.webhook_enabled,
            },
        }

    def clear_history(self) -> None:
        """Limpa histórico de alertas."""
        with self._lock:
            self._history.clear()
            self._aggregates.clear()
            self._last_sent.clear()

    def stop(self) -> None:
        """Para o gerenciador de alertas."""
        self._sender_running = False
        if self._sender_thread and self._sender_thread.is_alive():
            self._sender_thread.join(timeout=5.0)

File: src/monitor/alerts/alert-manager.ts
/**
 * Alert Manager - Sistema de Alertas
 *
 * Envia alertas para múltiplos canais:
 * - Console (log)
 * - Slack
 * - Discord
 * - Webhook customizado
 * - Telegram (via bot)
 */

import { EventEmitter } from 'events';
import { AlertConfig } from '../core/config';

export enum AlertLevel {
  DEBUG = 'debug',
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error',
  CRITICAL = 'critical',
}

export interface Alert {
  id: string;
  level: AlertLevel;
  title: string;
  message: string;
  source: string;
  timestamp: Date;
  tags?: Record<string, string>;
  metadata?: Record<string, any>;
  acknowledged: boolean;
  acknowledgedAt?: Date;
  acknowledgedBy?: string;
  sentTo: string[];
}

export interface AlertInput {
  level: AlertLevel;
  title: string;
  message: string;
  source: string;
  tags?: Record<string, string>;
  metadata?: Record<string, any>;
}

interface AlertAggregate {
  key: string;
  count: number;
  firstOccurrence: Date;
  lastOccurrence: Date;
  sample: Alert;
}

export class AlertManager extends EventEmitter {
  private history: Alert[] = [];
  private lastSent = new Map<string, number>();
  private aggregates = new Map<string, AlertAggregate>();
  private pendingAggregates = new Map<string, NodeJS.Timeout>();

  constructor(private readonly config: AlertConfig) {
    super();
  }

  /**
   * Envia um alerta.
   */
  send(input: AlertInput): Alert | null {
    if (!this.config.enabled) return null;

    const alert = this.createAlert(input);
    const key = this.getAlertKey(alert);

    // Verifica cooldown
    if (!this.shouldSend(key)) {
      this.aggregate(key, alert);
      return null;
    }

    // Registra envio
    this.lastSent.set(key, Date.now());
    this.history.push(alert);

    // Limita histórico
    if (this.history.length > 10000) {
      this.history = this.history.slice(-5000);
    }

    // Envia para canais
    this.sendToChannels(alert);

    // Emite evento
    this.emit('alert', alert);

    return alert;
  }

  /**
   * Cria objeto Alert.
   */
  private createAlert(input: AlertInput): Alert {
    return {
      id: this.generateId(),
      level: input.level,
      title: input.title,
      message: input.message,
      source: input.source,
      timestamp: new Date(),
      tags: input.tags,
      metadata: input.metadata,
      acknowledged: false,
      sentTo: [],
    };
  }

  /**
   * Gera ID único.
   */
  private generateId(): string {
    return `alert_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Gera chave para agregação/cooldown.
   */
  private getAlertKey(alert: Alert): string {
    return `${alert.level}:${alert.source}:${alert.title}`;
  }

  /**
   * Verifica se deve enviar (cooldown).
   */
  private shouldSend(key: string): boolean {
    const lastSentTime = this.lastSent.get(key) || 0;
    return Date.now() - lastSentTime >= this.config.cooldown;
  }

  /**
   * Agrega alertas similares.
   */
  private aggregate(key: string, alert: Alert): void {
    if (!this.config.aggregateAlerts) return;

    if (!this.aggregates.has(key)) {
      this.aggregates.set(key, {
        key,
        count: 1,
        firstOccurrence: alert.timestamp,
        lastOccurrence: alert.timestamp,
        sample: alert,
      });

      // Agenda envio do agregado
      const timer = setTimeout(() => {
        this.sendAggregated(key);
      }, this.config.cooldown);

      this.pendingAggregates.set(key, timer);
    } else {
      const agg = this.aggregates.get(key)!;
      agg.count++;
      agg.lastOccurrence = alert.timestamp;
    }
  }

  /**
   * Envia alerta agregado.
   */
  private sendAggregated(key: string): void {
    const agg = this.aggregates.get(key);
    if (!agg) return;

    const aggregatedAlert: Alert = {
      ...agg.sample,
      id: this.generateId(),
      message: `${agg.sample.message} (${agg.count} occurrences)`,
      metadata: {
        ...agg.sample.metadata,
        aggregated: true,
        count: agg.count,
        firstOccurrence: agg.firstOccurrence,
        lastOccurrence: agg.lastOccurrence,
      },
    };

    this.history.push(aggregatedAlert);
    this.sendToChannels(aggregatedAlert);
    this.emit('alert', aggregatedAlert);

    this.aggregates.delete(key);
    this.pendingAggregates.delete(key);
  }

  /**
   * Envia para todos os canais configurados.
   */
  private async sendToChannels(alert: Alert): Promise<void> {
    // Console
    this.logToConsole(alert);
    alert.sentTo.push('console');

    // Slack
    if (this.config.slackWebhook) {
      try {
        await this.sendToSlack(alert);
        alert.sentTo.push('slack');
      } catch (err) {
        console.error('Failed to send to Slack:', err);
      }
    }

    // Discord
    if (this.config.discordWebhook) {
      try {
        await this.sendToDiscord(alert);
        alert.sentTo.push('discord');
      } catch (err) {
        console.error('Failed to send to Discord:', err);
      }
    }

    // Custom Webhook
    if (this.config.customWebhook) {
      try {
        await this.sendToWebhook(alert);
        alert.sentTo.push('webhook');
      } catch (err) {
        console.error('Failed to send to webhook:', err);
      }
    }
  }

  /**
   * Log no console.
   */
  private logToConsole(alert: Alert): void {
    const levelColors: Record<string, string> = {
      debug: '\x1b[90m',
      info: '\x1b[36m',
      warning: '\x1b[33m',
      error: '\x1b[31m',
      critical: '\x1b[35m',
    };

    const reset = '\x1b[0m';
    const color = levelColors[alert.level] || reset;
    const timestamp = alert.timestamp.toISOString();

    console.log(
      `${color}[AURORA ${alert.level.toUpperCase()}]${reset} ${timestamp} | ${alert.source} | ${alert.title}: ${alert.message}`
    );
  }

  /**
   * Envia para Slack.
   */
  private async sendToSlack(alert: Alert): Promise<void> {
    const colorMap: Record<string, string> = {
      debug: '#808080',
      info: '#36a64f',
      warning: '#ffcc00',
      error: '#ff6600',
      critical: '#ff0000',
    };

    const payload = {
      attachments: [{
        color: colorMap[alert.level] || '#808080',
        title: `[${alert.level.toUpperCase()}] ${alert.title}`,
        text: alert.message,
        fields: [
          { title: 'Source', value: alert.source, short: true },
          { title: 'Time', value: alert.timestamp.toISOString(), short: true },
        ],
        footer: 'Aurora Monitor',
      }],
    };

    await fetch(this.config.slackWebhook!, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
  }

  /**
   * Envia para Discord.
   */
  private async sendToDiscord(alert: Alert): Promise<void> {
    const colorMap: Record<string, number> = {
      debug: 0x808080,
      info: 0x36a64f,
      warning: 0xffcc00,
      error: 0xff6600,
      critical: 0xff0000,
    };

    const payload = {
      embeds: [{
        color: colorMap[alert.level] || 0x808080,
        title: `[${alert.level.toUpperCase()}] ${alert.title}`,
        description: alert.message,
        fields: [
          { name: 'Source', value: alert.source, inline: true },
          { name: 'Time', value: alert.timestamp.toISOString(), inline: true },
        ],
        footer: { text: 'Aurora Monitor' },
      }],
    };

    await fetch(this.config.discordWebhook!, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
  }

  /**
   * Envia para webhook customizado.
   */
  private async sendToWebhook(alert: Alert): Promise<void> {
    await fetch(this.config.customWebhook!, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(alert),
    });
  }

  /**
   * Reconhece um alerta.
   */
  acknowledge(alertId: string, by: string = 'system'): boolean {
    const alert = this.history.find(a => a.id === alertId);
    if (!alert) return false;

    alert.acknowledged = true;
    alert.acknowledgedAt = new Date();
    alert.acknowledgedBy = by;

    this.emit('acknowledged', alert);
    return true;
  }

  /**
   * Retorna histórico.
   */
  getHistory(options?: {
    limit?: number;
    level?: AlertLevel;
    source?: string;
    since?: Date;
    acknowledged?: boolean;
  }): Alert[] {
    let alerts = [...this.history];

    if (options?.level) {
      const levelOrder = [AlertLevel.DEBUG, AlertLevel.INFO, AlertLevel.WARNING, AlertLevel.ERROR, AlertLevel.CRITICAL];
      const minIndex = levelOrder.indexOf(options.level);
      alerts = alerts.filter(a => levelOrder.indexOf(a.level) >= minIndex);
    }

    if (options?.source) {
      alerts = alerts.filter(a => a.source.includes(options.source!));
    }

    if (options?.since) {
      alerts = alerts.filter(a => a.timestamp >= options.since!);
    }

    if (options?.acknowledged !== undefined) {
      alerts = alerts.filter(a => a.acknowledged === options.acknowledged);
    }

    if (options?.limit) {
      alerts = alerts.slice(-options.limit);
    }

    return alerts;
  }

  /**
   * Retorna alertas não reconhecidos.
   */
  getUnacknowledged(level?: AlertLevel): Alert[] {
    return this.getHistory({ acknowledged: false, level });
  }

  /**
   * Retorna estatísticas.
   */
  getStats(): {
    total: number;
    unacknowledged: number;
    last24h: number;
    byLevel: Record<string, number>;
    bySource: Record<string, number>;
  } {
    const byLevel: Record<string, number> = {};
    const bySource: Record<string, number> = {};
    const cutoff = Date.now() - 24 * 60 * 60 * 1000;
    let last24h = 0;
    let unacknowledged = 0;

    for (const alert of this.history) {
      byLevel[alert.level] = (byLevel[alert.level] || 0) + 1;

      const sourceBase = alert.source.split('.')[0];
      bySource[sourceBase] = (bySource[sourceBase] || 0) + 1;

      if (alert.timestamp.getTime() > cutoff) last24h++;
      if (!alert.acknowledged) unacknowledged++;
    }

    return {
      total: this.history.length,
      unacknowledged,
      last24h,
      byLevel,
      bySource,
    };
  }

  /**
   * Limpa histórico.
   */
  clearHistory(): void {
    this.history = [];
    this.aggregates.clear();
    for (const timer of this.pendingAggregates.values()) {
      clearTimeout(timer);
    }
    this.pendingAggregates.clear();
    this.lastSent.clear();
  }
}
